### PR TITLE
Export cmake target

### PR DESCRIPTION
**Pull request by uilianries**
This PR includes a feature for exporting Zlib NG CMake targets in CMake files. This way, no only keeps an official naming when generating these files through package managers, but also provide an official support on it.

It considers too the cases where Zlib compat is active, so it follows the official ZLIB CMake module.

Some information about the output

----

Regular zlib-ng:

```
$ cmake --build build --target=install 
...
Install the project...
cmake -P cmake_install.cmake
-- Install configuration: "Release"
-- Installing: /tmp/install/lib/libz-ng.2.1.4.dylib
-- Installing: /tmp/install/lib/libz-ng.2.dylib
-- Installing: /tmp/install/lib/libz-ng.dylib
-- Installing: /tmp/install/lib/libz-ng.a
-- Installing: /tmp/install/include/zlib-ng.h
-- Installing: /tmp/install/include/zlib_name_mangling-ng.h
-- Installing: /tmp/install/include/zconf-ng.h
-- Installing: /tmp/install/lib/pkgconfig/zlib-ng.pc
-- Up-to-date: /tmp/install/lib/libz-ng.2.1.4.dylib
-- Up-to-date: /tmp/install/lib/libz-ng.2.dylib
-- Up-to-date: /tmp/install/lib/libz-ng.dylib
-- Installing: /tmp/install/lib/libz-ng.a
-- Installing: /tmp/install/lib/cmake/zlib-ng/zlib-ng.cmake
-- Installing: /tmp/install/lib/cmake/zlib-ng/zlib-ng-release.cmake
```

Where, both targets are created:

```
add_library(zlib-ng::zlib SHARED IMPORTED)
add_library(zlib-ng::zlibstatic STATIC IMPORTED)
```

----

With `ZLIB_COMPAT=ON`, which mimics the official CMake module:

https://cmake.org/cmake/help/latest/module/FindZLIB.html

```
$ cmake --build build --target=install 
...
Install the project...
cmake -P cmake_install.cmake
-- Install configuration: "Release"
-- Installing: /tmp/install/lib/libz.1.3.0.zlib-ng.dylib
-- Installing: /tmp/install/lib/libz.1.dylib
-- Installing: /tmp/install/lib/libz.dylib
-- Installing: /tmp/install/lib/libz.a
-- Installing: /tmp/install/include/zlib.h
-- Installing: /tmp/install/include/zlib_name_mangling.h
-- Installing: /tmp/install/include/zconf.h
-- Installing: /tmp/install/lib/pkgconfig/zlib.pc
-- Up-to-date: /tmp/install/lib/libz.1.3.0.zlib-ng.dylib
-- Up-to-date: /tmp/install/lib/libz.1.dylib
-- Up-to-date: /tmp/install/lib/libz.dylib
-- Up-to-date: /tmp/install/lib/libz.a
-- Installing: /tmp/install/lib/cmake/ZLIB/ZLIB.cmake
-- Installing: /tmp/install/lib/cmake/ZLIB/ZLIB-release.cmake
```

Where it provides both targets, but renames the shared one to UPPER-CASE:

```
add_library(ZLIB::ZLIB SHARED IMPORTED)
add_library(ZLIB::zlibstatic STATIC IMPORTED)
...
set_property(TARGET ZLIB::ZLIB APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(ZLIB::ZLIB PROPERTIES
  IMPORTED_LOCATION_RELEASE "${_IMPORT_PREFIX}/lib/libz.1.3.0.zlib-ng.dylib"
  IMPORTED_SONAME_RELEASE "@rpath/libz.1.dylib"
  )
```

In case only one is available (e.g. static), it will be renamed instead.

### CODE REVIEW

**Comment by Dead2** [comment]
@uilianries Could you please rebase this on top of current `develop`?
I am hoping the CI will now succeed. IIRC, this PR was made right in the middle of a couple PRs that temporarily broke the CI Package Checks.

**Comment by Dead2** [comment]
As far as the PR itself goes, it looks OK to me, but I honestly have no idea whether this is a good or bad way to do this, since I have never used CMake in this way myself.
It would be great if someone else would be able to provide a second set of eyes on this, either as a review or at least some justification for the need etc.

I am thinking it might be a good idea to provide a comment above line 1204 to explain what is happening here and why, to minimize the chance of this breaking later.
Also adding a CI check for it would be even better of course.

**Comment by uilianries** [comment]
> @uilianries Could you please rebase this on top of current `develop`? I am hoping the CI will now succeed. IIRC, this PR was made right in the middle of a couple PRs that temporarily broke the CI Package Checks.

Done, rebased.

**Comment by uilianries** [comment]
> I am thinking it might be a good idea to provide a comment above line 1204 to explain what is happening here and why, to minimize the chance of this breaking later.

Agreed, I just added a new comment in the CMakeLists.txt, please, check if it's clear enough.

**Comment by uilianries** [comment]
The current error on `Package Check / macOS Clang Symbol Prefix (pull_request) ` looks be expected, because this PR is installing new CMake files. I just need to know how to configure it in the CI to say: I'm installing xxx.cmake and want to check it as truth.

**Comment by Dead2** [comment]
@mtl1979 Am I right in thinking this is one of the cases where we need to update the reference git commit after merging the PR?

**Comment by mtl1979** [comment]
> @mtl1979 Am I right in thinking this is one of the cases where we need to update the reference git commit after merging the PR?

I don't think so... It's more about configure not creating the files at all... We might want to add an exception to the tests as those files are useless when building with configure.

```
diff --git a/test/pkgcheck.sh b/test/pkgcheck.sh
index 94a67a19..8464acf1 100644
--- a/test/pkgcheck.sh
+++ b/test/pkgcheck.sh
@@ -150,6 +150,9 @@ Darwin)
   ;;
 esac

+# Remove cmake target files
+find pkgtmp2 -type f -name '*.cmake' -exec rm '{}' \;
+
 # The ar on newer systems defaults to -D (i.e. deterministic),
 # but FreeBSD 12.1, Debian 8, and Ubuntu 14.04 seem to not do that.
 # I had trouble passing -D safely to the ar inside CMakeLists.txt,
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -1199,6 +1201,21 @@ if(NOT SKIP_INSTALL_HEADERS AND NOT SKIP_INSTALL_ALL)
...
+    # Generates ZLIB.cmake in case ZLIB_COMPAT=ON and always exports the CMake target ZLIB::ZLIB
+    # In case ZLIB_COMPAT=OFF, the CMake target and file follows zlib-ng naming convention
+    if (ZLIB_COMPAT)
+        if (TARGET zlib)
+            set_target_properties(zlib PROPERTIES EXPORT_NAME ZLIB)
+        else()
+            set_target_properties(zlibstatic PROPERTIES EXPORT_NAME ZLIB)
+        endif()
+    endif()
+    install(TARGETS ${ZLIB_INSTALL_LIBRARIES}
```
Comment (mtl1979) [inline CMakeLists.txt:1214]: This doesn't work with some cmake versions, missing `DESTINATION` parameter.

Comment (uilianries) [reply CMakeLists.txt:1214]: @mtl1979 good catch, indeed I only tested with newer version of cmake.

I tested now with cmake 3.6.3 and the PR was broken due the DESTINATION missing in the install(). I added it now and is working. Thank you!

Comment (mtl1979) [reply CMakeLists.txt:1214]: I mostly use cmake 3.10.2 on Ubuntu 18.04 for testing incompatibility issues... It's pretty much the most recent cmake version that still doesn't support all the basic features we use in testing zlib-ng. I used to have Ubuntu 16.04 with even older cmake version but I couldn't install any updates or new packages as the repositories were taken offline.

Comment (uilianries) [reply CMakeLists.txt:1214]: @mtl1979 I run a docker image with Ubuntu 16.04 + CMake 3.6.3 (installed via pip).

Comment (mtl1979) [reply CMakeLists.txt:1214]: I used to use Docker until it became obvious it's just waste of disk space... Now I just launch Ubuntu straight from disk images as my CPU can do that without additional software.

**Comment by mtl1979** [comment]
@Dead2 Can you approve the CI workflows, I don't have permission to do that...

**Comment by mtl1979** [comment]
The fails on Package Check CI runs were expected as the workaround in `test/pkgcheck.sh` hasn't been applied yet...

**Review by Dead2 — APPROVED** [review]
LGTM

**Pull request merged**